### PR TITLE
Added NLP: Basic text classification with JAX & Flax

### PR DESCRIPTION
**Pull request by vfdev-5**
cc @melissawm to review

### CODE REVIEW

**Comment by jakevdp** [comment]
Looks really great – thanks for this! I'll wait on a review from @melissawm before merging.

File: docs/JAX_basic_text_classification.md [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+    jupytext_version: 1.15.2
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Basic Text classification with JAX & FLAX
+
+In this tutorial we learn how to perform text classification from raw text data and train a basic 1D Convnet to perform sentiment analysis using JAX. This tutorial is originally inspired by ["Text classification from scratch with Keras"](https://keras.io/examples/nlp/text_classification_from_scratch/#build-a-model) tutorial.
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:17]:
```suggestion
In this tutorial we learn how to perform text classification from raw text data and train a basic 1D Convnet to perform sentiment analysis using JAX. This tutorial is originally inspired by ["Text classification from scratch with Keras"](https://keras.io/examples/nlp/text_classification_from_scratch/#build-a-model).
```

File: docs/JAX_basic_text_classification.md [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+We are using IMDB movie review dataset to classify the review to "positive" and "negative" classes. We implement from scratch a simple model using Flax, train it and compute metrics on test set.
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:19]:
```suggestion
We will use the IMDB movie review dataset to classify the review to "positive" and "negative" classes. We implement from scratch a simple model using Flax, train it and compute metrics on the test set.
```

File: docs/JAX_basic_text_classification.md [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+
++++
+
+## Setup
+
+We will be using the following packages in this tutorial:
+- [Tiktoken](https://github.com/openai/tiktoken) to tokenize the raw text
+- [Grain](https://github.com/google/grain) for efficient data loading and batching
+- [tqdm](https://tqdm.github.io/) for a progress bar in the training
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:28]: nit:

```suggestion
- [tqdm](https://tqdm.github.io/) for a progress bar in the training step
```

File: docs/JAX_basic_text_classification.md [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+!echo "Number of negative samples in test set:"
+!ls /tmp/data/imdb/aclImdb/test/neg | wc -l
+!echo "First 10 files with positive samples in train/test sets:"
+!ls /tmp/data/imdb/aclImdb/train/pos | head
+!ls /tmp/data/imdb/aclImdb/test/pos | head
+!echo "Display a single positive sample:"
+!cat /tmp/data/imdb/aclImdb/train/pos/6248_7.txt
+```
+
+Next, we create the dataset Python class to read samples from the disk, use [Tiktoken](https://github.com/openai/tiktoken) to encode raw text into tokens and use [Grain](https://github.com/google/grain) for efficient data loading and batching.
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:61]:
```suggestion
Next, we will:
- create the dataset Python class to read samples from the disk;
- use [Tiktoken](https://github.com/openai/tiktoken) to encode raw text into tokens; and
- use [Grain](https://github.com/google/grain) for efficient data loading and batching.
```

File: docs/JAX_basic_text_classification.md [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+root_path = Path("/tmp/data/imdb/aclImdb/")
+train_dataset = SentimentAnalysisDataset(root_path / "train")
+test_dataset = SentimentAnalysisDataset(root_path / "test")
+
+print("- Number of samples in train and test sets:", len(train_dataset), len(test_dataset))
+print("- First train sample:", train_dataset[0])
+print("- First test sample:", test_dataset[0])
+```
+
+Now, we can create string to tokens preprocessing transformation and set up data loaders. We are going to use the GPT-2 tokenizer via [Tiktoken](https://github.com/openai/tiktoken).
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:103]: This reads better to me, but it's only a suggestion:
```suggestion
Now, we can create a string-to-tokens preprocessing transformation and set up data loaders. We are going to use the GPT-2 tokenizer via [Tiktoken](https://github.com/openai/tiktoken).
```

File: docs/JAX_basic_text_classification.md [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+        # Cut to max length
+        encoded = encoded[:self.max_length]
+        # Pad with zeros if needed
+        encoded = np.array(encoded + [0] * (self.max_length - len(encoded)))
+        return {
+            "text": encoded,
+            "label": data["label"],
+        }
+```
+
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:132]: If you are using two separate code cells for these actions, I would keep the grain import in the next cell so they are "self-contained" (they are not really self-contained, but it makes for a better cognitive unit)

File: docs/JAX_basic_text_classification.md [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+train_batch = next(iter(train_loader))
+```
+
+```{code-cell} ipython3
+print("Train encoded text batch info:", type(train_batch["text"]), train_batch["text"].shape, train_batch["text"].dtype)
+print("Train labels batch info:", type(train_batch["label"]), train_batch["label"].shape, train_batch["label"].dtype)
+```
+
+```{code-cell} ipython3
+print("Batch data:", train_batch["text"][:2, :12], train_batch["label"][:2])
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:193]: Can you leave a tiny note about why we are printing this, and what we should expect to see?

File: docs/JAX_basic_text_classification.md [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+```
+
+## Model for text classification
+
+We choose a simple 1D convnet to classify the text. The first layer of the model transforms input tokens into float features using an embedding layer (`nnx.Embed`), then they are encoded further with convolutions. Finally, we classify encoded features using a fully-connected layers.
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:198]:
```suggestion
We choose a simple 1D convnet to classify the text. The first layer of the model transforms input tokens into float features using an embedding layer (`nnx.Embed`), then they are encoded further with convolutions. Finally, we classify encoded features using fully-connected layers.
```

File: docs/JAX_basic_text_classification.md [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+        x = self.dropout(x)
+        x = self.fc2(x)  # x.shape: (N, 2)
+
+        return x
+
+
+x = jnp.ones((4, max_length), dtype="int32")
+module = TextConvNet(vocab_size)
+y = module(x)
+print(y.shape)
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:273]: Why is this shape informative here?

File: docs/JAX_basic_text_classification.md [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+
+    for metric, value in eval_metrics.compute().items():
+        eval_metrics_history[f'test_{metric}'].append(value)
+
+    print(f"[test] epoch: {epoch + 1}/{num_epochs}")
+    print(f"- total loss: {eval_metrics_history['test_loss'][-1]:0.4f}")
+    print(f"- Accuracy: {eval_metrics_history['test_accuracy'][-1]:0.4f}")
+```
+
+Let us now start the training
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:404]: Just a suggestion:
```suggestion
Now, we can start the training. Note that this step can take quite a while to run, depending on the number of epochs you are using. 
```

File: docs/JAX_basic_text_classification.md [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+
+```{code-cell} ipython3
+for epoch in range(num_epochs):
+    train_one_epoch(epoch)
+    evaluate_model(epoch)
+```
+
+Let us visualize collected metrics
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:412]:
```suggestion
Let's visualize the collected metrics:
```

File: docs/JAX_basic_text_classification.md [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+preds = model(processed_data["text"][None, :])
+preds, preds.argmax(axis=-1)
+```
+
+```{code-cell} ipython3
+print("Text:\n", data["text"])
+print("Expected label:", data["label"])
+print("Predicted label:", preds.argmax(axis=-1).item())
+```
+
```
Comment (melissawm) [inline docs/JAX_basic_text_classification.md:452]: Is it possible to include a quick summary/interpretation of what we are seeing in these labels?

**Review by melissawm — COMMENTED** [review]
Left a couple of optional suggestions, some to improve the text flow and some to include more insight into results/prints of the results. Cheers!

File: docs/JAX_basic_text_classification.md (hunk 9 above)
Comment (vfdev-5) [reply docs/JAX_basic_text_classification.md:273]: To ensure that we output expected shape: (N, num_classes), updated to
```
print("Prediction shape (N, num_classes): ", y.shape)
```

File: docs/JAX_basic_text_classification.md (hunk 10 above)
Comment (vfdev-5) [reply docs/JAX_basic_text_classification.md:404]: Well, this is not exactly true, in this case, whole training takes 2 minutes. The model and the dataset are rather small.

File: docs/JAX_basic_text_classification.md (hunk 3 above)
Comment (melissawm) [reply docs/JAX_basic_text_classification.md:28]: Just ignore then, thanks :)

File: docs/JAX_basic_text_classification.md (hunk 10 above)
Comment (melissawm) [reply docs/JAX_basic_text_classification.md:404]: Interesting! For me it took something like 40 mins in colab.

File: docs/JAX_basic_text_classification.md [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.15.2
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Basic Text classification with JAX & FLAX
```
Comment (jakevdp) [inline docs/JAX_basic_text_classification.md:15]: Let's say "1D Convnet for basic text classification"

File: docs/JAX_basic_text_classification.md [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+print("")
+print(f"- Expected review sentiment: {'positive' if data['label'] == 0 else 'negative'}")
+print(f"- Predicted review sentiment: {'positive' if pred_label == 0 else 'negative'}, confidence: {confidence[0, pred_label]:.3f}")
+```
+
+## Further reading
+
+- Model checkpointing and exporting using [Orbax](https://orbax.readthedocs.io/en/latest/)
+
+```{code-cell} ipython3
```
Comment (jakevdp) [inline docs/JAX_basic_text_classification.md:476]: Remove empty cell

File: docs/JAX_basic_text_classification.md [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+
+print("- Text:\n", data["text"])
+print("")
+print(f"- Expected review sentiment: {'positive' if data['label'] == 0 else 'negative'}")
+print(f"- Predicted review sentiment: {'positive' if pred_label == 0 else 'negative'}, confidence: {confidence[0, pred_label]:.3f}")
+```
+
+## Further reading
+
+- Model checkpointing and exporting using [Orbax](https://orbax.readthedocs.io/en/latest/)
```
Comment (jakevdp) [inline docs/JAX_basic_text_classification.md:474]: It would be nice to have some sort of conclusion text here – summarize what we did and what the next steps may be for someone interested in going deeper.

**Review by jakevdp — COMMENTED** [review]
Looks great, thanks!

File: docs/JAX_basic_text_classification.md (hunk 15 above)
Comment (vfdev-5) [reply docs/JAX_basic_text_classification.md:474]: Yes, I wanted to add links on other NLP tutorials doing text classification with transformers once we have them.
I'll add a short summary on what we did.

**Review by trallard — APPROVED** [review]
This seems all ready to merge after a couple of rounds of reviews cc/ @jakevdp 🚀

**Review by jakevdp — APPROVED** [review]
